lib/galaxy/objectstore/irods.py

```diff
@@ -0,0 +1,710 @@
+    """
+    Object store that stores objects as data objects in an iRODS collections. A local cache 
+    exists that is used as an intermediate location for files between Galaxy and iRODS.
+    """
+    
+    store_type = 'irods'
+    
+    #def _get_bucket(self, bucket_name):
+    #def _key_exists(self, rel_path):
+
+    def __init__(self, config, config_dict):
+        super(IRODSObjectStore, self).__init__(config, config_dict)
+
+        auth_dict = config_dict['auth']
+        self.username = auth_dict.get('username')
+        self.password = auth_dict.get('password')
+
+        resource_dict = config_dict['resource']
+        self.resource = resource_dict.get('name')
+
+        zone_dict = config_dict['zone']
+        self.zone = zone_dict.get('name')
+
+        connection_dict = config_dict['connection']
+        self.host = connection_dict.get('host')
+        self.port = connection_dict.get('port')
+
+        cache_dict = config_dict['cache']
+        self.cache_size = cache_dict.get('size', -1)
+        self.staging_path = cache_dict.get('path') or self.config.object_store_cache_path
+
+        extra_dirs = dict((e['type'], e['path']) for e in config_dict.get('extra_dirs', []))
+        self.extra_dirs.update(extra_dirs)
+
+        log.debug("Object cache dir: %s", self.staging_path)
+        log.debug("Job work dir: %s", self.extra_dirs['job_work'])
+
+        self._initialize()
+        
+    def __del__(self):
+        self.session.cleanup()
+
+    def _initialize(self):
+        if irods is None:
+            raise Exception(IRODS_IMPORT_MESSAGE)
+
+        self.home =  "/" + self.zone + "/home/" + self.username
+
+        self.session = self._configure_connection(host=self.host, port=self.port, user=self.username, password=self.password, zone=self.zone)
+        # Clean cache only if value is set in galaxy.ini
+        if self.cache_size != -1:
+            # Convert GBs to bytes for comparison
+            self.cache_size = self.cache_size * 1073741824
+            # Helper for interruptable sleep
+            self.sleeper = Sleeper()
+            self.cache_monitor_thread = threading.Thread(target=self.__cache_monitor)
+            self.cache_monitor_thread.start()
+            log.info("Cache cleaner manager started")
+
+    def _configure_connection(self, host='localhost', port='1247', user='rods', password='rods', zone='tempZone'):
+        with iRODSSession(host=host, port=port, user=user, password=password, zone=zone) as session:
+            # Throws NetworkException if connection fails  
+            session.pool.get_connection()
+            return session
+
+    @classmethod
+    def parse_xml(cls, config_xml):
+        return parse_config_xml(config_xml)
+
+    def to_dict(self):
+        as_dict = super(IRODSObjectStore, self).to_dict()
+        as_dict.update(self._config_to_dict())
+        return as_dict
+
+    def __cache_monitor(self):
+        time.sleep(2)  # Wait for things to load before starting the monitor
```

Comment:
What things are this ? Can we have a smarter check here ?

Reply:
Same thing here (I borrowed the caching code directly from s3.py. It works).

Reply:
This code was deleted per our discussion.